doc/internals.rst

```diff
@@ -129,9 +129,77 @@ Back in an interactive IPython session, we can use these properties:
 
 The intent here is that libraries that extend xarray could add such an accessor
 to implement subclass specific functionality rather than using actual subclasses
-or patching in a large number of domain specific methods. For further reading
-on ways to write new accessors and the philosophy behind the approach, see
-:issue:`1080`.
+or patching in a large number of domain specific methods.
+
+Parametrizing an accessor is possible by defining ``__call__``. For
+example, we could use this to reimplement the API of the
+:py:func:`DataArray.weighted` function:
+
+.. ipython::
+    :okwarning:
+
+    In [1]: @xr.register_dataarray_accessor("weighted")
+       ...: class WeightedAccessor:
+       ...:     def __init__(self, xarray_obj):
+       ...:         self._obj = xarray_obj
+       ...:         self._weights = np.ones_like(xarray_obj) / xarray_obj.size
+       ...:
+       ...:     def __call__(self, weights):
+       ...:         self._weights = weights
+       ...:         return self
+       ...:
+       ...:     def sum(self, *args, **kwargs):
+       ...:         return np.sum(self._obj * self._weights, *args, **kwargs)
+
+    In [2]: da = xr.DataArray(data=np.linspace(1, 2, 10), dims="x")
+       ...: weights = xr.DataArray(
+       ...:     np.array([0.25, 0, 0, 0, 0.25, 0, 0.25, 0, 0, 0.25]),
+       ...:     dims="x",
+       ...: )
+
+    In [3]: da.weighted.sum()
+
+    In [4]: da.weighted(weights).sum()
+
+If we want to require the parameter, the easiest way to do so is using
+a wrapper function:
+
+.. ipython::
+    :okwarning:
+    :okexcept:
+
+    In [1]: class Weighted:
+       ...:     def __init__(self, obj, weights):
+       ...:         self._obj = obj
+       ...:         self._weights = weights
+       ...:
+       ...:     def sum(self, *args, **kwargs):
+       ...:         return np.sum(self._obj * self._weights, *args, **kwargs)
+
+    In [2]: @xr.register_dataarray_accessor("weighted")
+       ...: def weighted(obj):
+       ...:     def wrapped(weights):
+       ...:         return Weighted(obj, weights)
+       ...:     return wrapped
```

Comment:
Using `partial` from functools could look a little cleaner here?
```suggestion
       ...:     return partial(Weighted, obj)
```

I guess the downside is that introspection on the argument args/signature probably wouldn't work quite as well.

Reply:
This looks somewhat like a decorator, but I'm not really sure if that's good or bad. The missing introspection is not that much of an issue: contrary to properties, autocompleters usually don't complete the attributes of return values of function calls. However, the signature and docstring is lost on `partial` objects, so `help()` becomes useless. Also, the error message when forgetting to call the accessor could be better (a `AttributeError` complaining about `function` or the `partial` object not having the attribute).

One way to work around that could be to extend the accessor registration functions to automatically (or only if a keyword-only argument was set to `True`) use a wrapper class if the accessor defines `__call__` or some sort of hook method. That wrapper would then do the same as the function, but provide more helpful error messages.

I'm thinking of something like
```python
class ParametrizedAccessorWrapper:
    def __init__(self, accessor_cls):
        self._accessor_cls = accessor_cls

    def __call__(self, *args, **kwargs):
        return self._accessor_cls(*args, **kwargs)

    def __getattr__(self, name):
        raise AttributeError(
            f"cannot access attribute {name!r} without parametrizing the accessor"
        )
```